DllImportGenerator/DllImportGenerator/Marshalling/MarshallingGenerator.cs

```diff
@@ -96,51 +137,54 @@ public static bool TryCreate(TypePositionInfo info, StubCodeContext context, out
                     or { ManagedType: { SpecialType: SpecialType.System_Double }, MarshallingAttributeInfo: null or MarshalAsInfo { UnmanagedType: UnmanagedType.R8 } }:
-                    generator = Blittable;
-                    return true;
+                    return Blittable;
 
                 case { ManagedType: { SpecialType: SpecialType.System_Boolean }, MarshallingAttributeInfo: null }:
-                    generator = WinBool; // [Compat] Matching the default for the built-in runtime marshallers.
-                    return true;
+                    return WinBool; // [Compat] Matching the default for the built-in runtime marshallers.
                 case { ManagedType: { SpecialType: SpecialType.System_Boolean }, MarshallingAttributeInfo: MarshalAsInfo { UnmanagedType: UnmanagedType.I1 or UnmanagedType.U1 } }:
-                    generator = ByteBool;
-                    return true;
+                    return ByteBool;
                 case { ManagedType: { SpecialType: SpecialType.System_Boolean }, MarshallingAttributeInfo: MarshalAsInfo { UnmanagedType: UnmanagedType.I4 or UnmanagedType.U4 or UnmanagedType.Bool } }:
-                    generator = WinBool;
-                    return true;
+                    return WinBool;
                 case { ManagedType: { SpecialType: SpecialType.System_Boolean }, MarshallingAttributeInfo: MarshalAsInfo { UnmanagedType: UnmanagedType.VariantBool } }:
-                    generator = VariantBool;
-                    return true;
+                    return VariantBool;
+
+                case { ManagedType: { SpecialType: SpecialType.System_Char }, MarshallingAttributeInfo: null }:
+                    return Utf16Char; // [Compat] Default marshalling is UTF-16.
+                case { ManagedType: { SpecialType: SpecialType.System_Char }, MarshallingAttributeInfo: MarshallingInfoStringSupport { CharEncoding: CharEncoding.Utf8 } }:
+                    throw new MarshallingNotSupportedException(info, context) // [Compat] See conversion from CharSet.Ansi to UTF-8.
+                    {
+                        NotSupportedDetails = Resources.MarshallingCharAsCharSetAnsiNotSupported
+                    };
+                case { ManagedType: { SpecialType: SpecialType.System_Char }, MarshallingAttributeInfo: MarshallingInfoStringSupport { CharEncoding: CharEncoding.PlatformDefined } }:
+                    throw new MarshallingNotSupportedException(info, context) // [Compat] See conversion of CharSet.Auto.
+                    {
+                        NotSupportedDetails = Resources.MarshallingCharAsCharSetAutoNotSupported
+                    };
+                case { ManagedType: { SpecialType: SpecialType.System_Char }, MarshallingAttributeInfo: MarshallingInfoStringSupport { CharEncoding: CharEncoding.Utf16 } }:
+                    return Utf16Char;
```

Comment:
If the `UnmanagedType` is explicitly set, shouldn't that take precedence over the `CharSet` such that `I1`/`U1` -> UTF-8 and `I2`/`U2` -> UTF-16?

Reply:
Also, do we care to explicitly not support / error if the `UnmanagedType` is not one of those?

Reply:
>Also, do we care to explicitly not support / error if the UnmanagedType is not one of those?

I don't think so. It indicates the `MarshalAs` is invalid for the parameter so I think that works.